juju/api.go

```diff
@@ -381,10 +387,57 @@ func updateControllerDetailsFromLogin(
 	} else if reordered {
 		logger.Tracef("API endpoints reordered from %v to %v", details.APIEndpoints, newDetails.APIEndpoints)
 	}
-	err = store.UpdateController(controllerName, *newDetails)
+	err := store.UpdateController(controllerName, *newDetails)
 	return errors.Trace(err)
 }
 
+func trimScheme(urls []*url.URL) []string {
+	res := make([]string, len(urls))
+	for i, url := range urls {
+		res[i] = url.Host + url.RequestURI()
+	}
+	return res
+}
+
+// makeUsableAddresses returns a list of controller addresses
+// in a format appropriate for persisting in the Juju client store.
+// The addresses will be a URL but omit any scheme i.e. <domain>:<port>/<path>
+// The addresses are filtered to only those unique and usable.
+// Finally, the params.DNSCache will be updated.
+func makeUsableAddresses(params *UpdateControllerParams) []string {
+	// Use a defaultScheme local to this function for comparisons
+	// between the list of HostPorts and the connectedURL.
+	defaultScheme := "scheme"
+
+	// HostPorts don't have a scheme, so for comparisons use the default.
+	addresses := usableHostPorts(params.CurrentHostPorts).CanonicalURLs(defaultScheme)
+
+	// Ignore the currently connected address if there is no current
+	// connection (during bootstrap) or if the connection is proxied.
+	if params.CurrentConnection == nil || params.CurrentConnection.Proxied {
+		return trimScheme(addresses)
+	}
+
+	// Make a copy to avoid modifying the provided address.
+	connectedUrl := &url.URL{}
+	*connectedUrl = *params.CurrentConnection.Address
+	connectedUrl.Scheme = defaultScheme
```

Comment:
Wait a minute. If we are using a dummy scheme to compare for equivalence, why not manipulate the stringified version of this and skip the slice transformation altogether?

Reply:
Yes good catch, I've cleaned up this function to do that and be less confusing.